### ISSUE TITLE
odo dev part 2 (cleanup)

### ISSUE THREAD

**Issue by kadel** [redhat-developer/odo#5485]
## Acceptance Criteria
- [x] Before the command ends it should delete all **inner loop** resources that it created on the cluster.
   - [x] There are different ways how the command can end (for example SIGQUIT, SIGHUP, SIGINT etc...) Program should catch all those signals and do cleanup before exiting.
   - [x] if the command is interrupted in the middle of some other operation (for example file syncing, k8s resource creation etc...) it should complete what is is doing, and when it gets to the point when `odo dev` is waiting or files to change start the cleanup procedure
- [x] The last line of the output needs to always inform user that ctrl+c will end the command and that it will result in deleting all resources from the cluster
- [ ] at least one test with devfile that creates storage, k8s. networking/Service (endpoint), k8s component.  (moved to separate issue https://github.com/redhat-developer/odo/issues/5621)
- [ ] verify that this work on Windows (moved to separate issue https://github.com/redhat-developer/odo/issues/5621)
- [x] for now this won't wait for resources to get deleted, it can reuse current `odo delete component` logic. It should show warning that it might take some time for resources to get deleted

## blocked on
- https://github.com/redhat-developer/odo/issues/5484

**Comment by dharmit** [comment]
@kadel @feloy so far, we always assign the Deployment's UUID as `ownerReferences` to every inner-loop resource created by `odo push` and hence by `odo dev`. I can't think of doing more than deleting this main Kubernetes Deployment and letting its garbage collector do the rest of the cleanup for odo.

Do you have any other scenarios in mind that I might be missing here?

**Comment by feloy** [comment]
> @kadel @feloy so far, we always assign the Deployment's UUID as `ownerReferences` to every inner-loop resource created by `odo push` and hence by `odo dev`. I can't think of doing more than deleting this main Kubernetes Deployment and letting its garbage collector do the rest of the cleanup for odo.
> 
> Do you have any other scenarios in mind that I might be missing here?

Good point, as all resources are linked to the parent Deployment with ownerReferences, the only resource to memorize is the Deployment one.

**Comment by kadel** [comment]
> @kadel @feloy so far, we always assign the Deployment's UUID as `ownerReferences` to every inner-loop resource created by `odo push` and hence by `odo dev`. I can't think of doing more than deleting this main Kubernetes Deployment and letting its garbage collector do the rest of the cleanup for odo.

If `ownersReference` is always used for all created resources then we can use it. But we need to make sure that we never forget to add `ownersReference`

**Comment by dharmit** [comment]
> Good point, as all resources are linked to the parent Deployment with ownerReferences, the only resource to memorize is the Deployment one.

IMO, we don't need to memorize it. We can always create a label selector based on the name in the `devfile.yaml` to get a matching Deployment from the cluster.


> If `ownersReference` is always used for all created resources then we can use it. But we need to make sure that we never forget to add `ownersReference`

Yes. And we are already doing this upon `odo push`. That's what `odo delete` has been relying on. But I'll check the code to be sure.

**Comment by dharmit** [comment]
>   * [ ]  if the command is interrupted in the middle of some other operation (for example file syncing, k8s resource creation etc...) it should complete what is is doing, and when it gets to the point when `odo dev` is waiting or files to change start the cleanup procedure

What's the rationale behind this? My understanding is that the user would hit `Ctrl+C` in the middle of `odo dev` to stop deployment of their component, maybe because they forgot something or want to change something. Why do we want to persist with deploying this component and wait till whatever time it takes to do so before cleaning up the resources, if any?

**Comment by dharmit** [comment]
I was playing with the code I found on this [gist](https://gist.github.com/fracasula/b579d52daf15426e58aa133d0340ccb0). I only added a sleep of 10 seconds and a `waitGroup.Add(1)` to the code compared to the original code:

```go
package main

// Here's a simple example to show how to properly terminate multiple go routines by using a context.
// Thanks to the WaitGroup we'll be able to end all go routines gracefully before the main function ends.

import (
	"context"
	"fmt"
	"math/rand"
	"os"
	"sync"
	"time"
)

func init() {
	rand.Seed(time.Now().UnixNano())
}

// it prints all values pushed into "ch" ("ch" here is read only)
func reader(ctx context.Context, wg *sync.WaitGroup, ch <-chan int) {
	defer wg.Done() // decrements the WaitGroup counter by one when the function returns

	for {
		select {
		case <-ctx.Done(): // Done returns a channel that's closed when work done on behalf of this context is canceled
			fmt.Println("Exiting from reading go routine")
			return
		case v, ok := <-ch:
			if !ok {
				fmt.Println("Channel has been closed")
				return
			}

			fmt.Println(v)
		}
	}
}

// it writes a random integer into "ch" every second ("ch" here is write only)
func writer(ctx context.Context, wg *sync.WaitGroup, ch chan<- int) {
	defer wg.Done() // decrements the WaitGroup counter by one when the function returns

	for {
		select {
		case <-ctx.Done(): // Done returns a channel that's closed when work done on behalf of this context is canceled
			fmt.Println("Exiting from writing go routine")
			return
		case ch <- rand.Intn(100): // pushes a random integer from 0 to 100 into the channel
			time.Sleep(1 * time.Second) // sleeps one second
		}
	}
}

func main() {
	channel := make(chan int)
	defer close(channel)

	// a WaitGroup waits for a collection of goroutines to finish, pass this by address
	waitGroup := sync.WaitGroup{}

	// context.WithCancel returns a copy of parent with a new Done channel.
	// The returned context's Done channel is closed when the returned cancel function is called or when the parent
	// context's Done channel is closed, whichever happens first.
	ctx, cancel := context.WithCancel(context.Background())

	waitGroup.Add(2)                    // adds delta, if the counter becomes zero, all goroutines blocked on Wait are released
	go reader(ctx, &waitGroup, channel) // go routine that prints all values pushed into "channel"
	go writer(ctx, &waitGroup, channel) // go routine that writes a random integer into "channel" every second

	// go routine that listens for an Enter keystroke to terminate the program
	waitGroup.Add(1)
	go func() {
		os.Stdin.Read(make([]byte, 1)) // wait for Enter keystroke
		cancel()                       // cancel the associated context
		fmt.Println("sleeping 10 seconds")
		time.Sleep(10 * time.Second)
		waitGroup.Done()
	}()

	waitGroup.Wait() // it blocks until the WaitGroup counter is zero

	fmt.Println("Exiting from main")
}
```

* With context being shared among the go routines, the `reader` and `writer` go routine are stopped when we call `cancel()`.
* When we don't use the context, `reader` and `writer` go routines still do their thing after the user hits return/Enter key.

For our use case, IMO, we need to share the context so that watch functionality doesn't trigger a push when we are trying to do cleanup.

**Comment by feloy** [comment]
Great code example, thanks. 
We need to be sure that we have in the go routines some point of synchronization that we can use to catch the Done(), I mean we need to have a `select` somewhere to catch it

**Comment by dharmit** [comment]
> We need to be sure that we have in the go routines some point of synchronization that we can use to catch the Done(), I mean we need to have a `select` somewhere to catch it. Is it the case?

Yes. I'm wondering if we can put it in this go routine in `watch.go`: https://github.com/redhat-developer/odo/blob/d48a607e9a76112ae1df5d883d714843722f609b/pkg/watch/watch.go#L196

**Comment by dharmit** [comment]
Removing needs-info because I think I have what I need to start on this.

/remove-triage needs-informatoin
